### PR TITLE
Fix Stratification of Non Atoms

### LINKED ISSUE #560: Possible bug related to stratification of non-atoms
Consider the program:

```
        rel A(x: Int)
        rel B(x: Int)
        rel R(x: Int)

        A(1).
        A(2).
        A(3).

        B(3).

        R(x) :- A(x), not B(x). // rule1

        false :- not R(1).  // rule2
        false :- not R(2).  // rule3
        false :- R(3).  // rule4
```

Here I would expect rule2-4 to be put into stratum 3, but they are put into stratum 0.

### PR DESCRIPTION
**Pull request by jmittert**
The stratifier was ignoring non atoms. It now handles them properly. This should fix #560.

I'm not sure how to test this though. Is there a way to have a test stop after a given phase and make assertions?

I can show that in the debugger, running the sample program in #560 that rules 2-4 are put in the 3rd Stratum.
image

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala [hunk 1]
```diff
@@ -62,14 +62,33 @@ object Stratifier extends Phase[TypedAst.Root, TypedAst.Root] {
   }
 
   /**
+    * A small case class to allow us to abstract over the possible heads in a
+    * constraint
+    */
+  trait ConstrHead
+  object ConstrHead {
+    case class Atom(symbol: Symbol.TableSym) extends ConstrHead
+    case class True() extends ConstrHead
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala]: `case object`.

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala [hunk 2]
```diff
@@ -62,14 +62,33 @@ object Stratifier extends Phase[TypedAst.Root, TypedAst.Root] {
...
+    case class False() extends ConstrHead
+  }
+
+  def makeConstrHead(constr: TypedAst.Constraint): ConstrHead = constr.head match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala]: Documentation

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala]: Also, in other places, we list true and false before Atom.

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala [hunk 3]
```diff
@@ -62,14 +62,33 @@ object Stratifier extends Phase[TypedAst.Root, TypedAst.Root] {
   }
 
   /**
+    * A small case class to allow us to abstract over the possible heads in a
+    * constraint
+    */
+  trait ConstrHead
+  object ConstrHead {
+    case class Atom(symbol: Symbol.TableSym) extends ConstrHead
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala]: `symbol -> sym` per naming convention.

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala [hunk 4]
```diff
@@ -62,14 +62,33 @@ object Stratifier extends Phase[TypedAst.Root, TypedAst.Root] {
   }
 
   /**
+    * A small case class to allow us to abstract over the possible heads in a
+    * constraint
+    */
+  trait ConstrHead
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala]: sealed trait

**Comment by magnus-madsen** [comment]
I added some minor stylistic issues, after those are fixed, we can merge it in.

**Comment by jmittert** [comment]
I've updated the PR for your feedback.

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**